kv/kvserver: TestReplicaCircuitBreaker_RangeFeed failed

kv/kvserver.TestReplicaCircuitBreaker_RangeFeed [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4424189&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4424189&tab=artifacts#/) on master @ [1a89121d67fa02d5c674766ea01d3c78455fd28b](https://github.com/cockroachdb/cockroach/commits/1a89121d67fa02d5c674766ea01d3c78455fd28b):


```
=== RUN   TestReplicaCircuitBreaker_RangeFeed
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaCircuitBreaker_RangeFeed1722575909
    test_log_scope.go:80: use -show-logs to present logs inline
    client_replica_circuit_breaker_test.go:426: val:<key:"\372" value:<raw_bytes:"+\304\231\334\003hello" timestamp:<wall_time:1645451313227447418 > > prev_value:<timestamp:<> > > 
    replica_raft.go:1111: 1 RequestLease: using slow replication threshold 10ms
    replica_raft.go:1111: 1 Probe: using slow replication threshold 10ms
    client_replica_circuit_breaker_test.go:426: val:<key:"\372" value:<raw_bytes:"+\304\231\334\003hello" timestamp:<wall_time:1645451316137942899 > > prev_value:<timestamp:<> > > 
    client_replica_circuit_breaker_test.go:424: context deadline exceeded
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaCircuitBreaker_RangeFeed1722575909
--- FAIL: TestReplicaCircuitBreaker_RangeFeed (50.80s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCircuitBreaker_RangeFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-13304

--- a/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
+++ b/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
@@ -427,7 +427,6 @@ func (s *dummyStream) Send(ev *roachpb.RangeFeedEvent) error {
 // goes as far as actually losing quorum to verify this end-to-end.
 func TestReplicaCircuitBreaker_RangeFeed(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 76856, "flaky test")
 	defer log.Scope(t).Close(t)
 	tc := setupCircuitBreakerTest(t)
 	ctx := context.Background()
@@ -479,7 +478,19 @@ func TestReplicaCircuitBreaker_RangeFeed(t *testing.T) {
 	undo()
 	tc.SetSlowThreshold(10 * time.Millisecond)
 	tc.StopServer(n2)
-	tc.RequireIsBreakerOpen(t, tc.Write(n1))
+	// NB: this is essentially always a breaker error but we can also get
+	// NotLeaseholderError(heartbeat failed on epoch increment), so we retry
+	// until we see the breaker error.
+	testutils.SucceedsSoon(t, func() error {
+		err := tc.Write(n1)
+		if err == nil {
+			return errors.New("no error returned")
+		}
+		if !tc.IsBreakerOpen(err) {
+			return err
+		}
+		return nil
+	})
 
 	// Start another stream during the "outage" to make sure it isn't rejected by
 	// the breaker.
@@ -501,7 +512,9 @@ func TestReplicaCircuitBreaker_RangeFeed(t *testing.T) {
 	// return a circuit breaker error, but in theory it could also never have
 	// tried to acquire a lease, in which case it might return a value as well.
 	if err := readOneVal(ctx, stream2, testutils.DefaultSucceedsSoonDuration); err != nil {
-		tc.RequireIsBreakerOpen(t, err)
+		if !tc.IsBreakerOpen(err) && !errors.HasType(err, (*roachpb.RangeFeedRetryError)(nil)) {
+			t.Fatalf("%+v", err)
+		}
 	}
 }
 
@@ -930,14 +943,23 @@ func (cbt *circuitBreakerTest) sendViaDistSender(
 	return pErr.GoError()
 }
 
-func (*circuitBreakerTest) RequireIsBreakerOpen(t *testing.T, err error) {
+func (*circuitBreakerTest) IsBreakerOpen(err error) bool {
+	// NB: we will see AmbiguousResultError here when proposals are inflight while
+	// the breaker trips. These are wrapping errors, so the assertions below will
+	// look through them.
+	if !errors.Is(err, circuit.ErrBreakerOpen) {
+		return false
+	}
+	return errors.HasType(err, (*roachpb.ReplicaUnavailableError)(nil))
+}
+
+func (cbt *circuitBreakerTest) RequireIsBreakerOpen(t *testing.T, err error) {
 	t.Helper()
 	t.Log(err)
 	// NB: we will see AmbiguousResultError here when proposals are inflight while
 	// the breaker trips. These are wrapping errors, so the assertions below will
 	// look through them.
-	require.True(t, errors.Is(err, circuit.ErrBreakerOpen), "%+v", err)
-	require.True(t, errors.HasType(err, (*roachpb.ReplicaUnavailableError)(nil)), "%+v", err)
+	require.True(t, cbt.IsBreakerOpen(err), "%+v", err)
 }
 
 func (*circuitBreakerTest) RequireIsNotLeaseholderError(t *testing.T, err error) {

